File: frontend/src/utils/pathUtils.ts
/**
 * Utilidades para conversión de rutas entre formatos Linux y Windows.
 *
 * Este módulo proporciona funciones para convertir rutas entre el formato Linux
 * usado en el servidor y el formato Windows UNC necesario para que los usuarios
 * puedan acceder a los archivos desde sus equipos Windows.
 *
 * Equivalente al path_utils.py del backend.
 */

/**
 * Convierte una ruta de Linux a formato Windows UNC.
 *
 * Conversión:
 *   FROM: /mnt/repositorio/2510SP/...
 *   TO:   \\repositorio\DirGesCat\2510SP\...
 *
 * @param linuxPath - Ruta en formato Linux (ej: /mnt/repositorio/2510SP/...)
 * @returns Ruta en formato Windows UNC o el valor original si no se puede convertir
 *
 * @example
 * linuxToWindowsPath('/mnt/repositorio/2510SP/H_Informacion_Consulta')
 * // => '\\\\repositorio\\DirGesCat\\2510SP\\H_Informacion_Consulta'
 *
 * linuxToWindowsPath('Sin ruta')
 * // => 'Sin ruta'
 *
 * linuxToWindowsPath(null)
 * // => null
 */
export function linuxToWindowsPath(linuxPath: string | null | undefined): string | null | undefined {
  // Si es null, undefined o vacío, retornar tal cual
  if (!linuxPath) {
    return linuxPath
  }

  // Si no es una cadena válida, retornar tal cual
  if (typeof linuxPath !== 'string') {
    return linuxPath
  }

  // Normalizar la ruta para procesar
  const rutaNormalizada = linuxPath.trim()

  // Caso 1: Ruta Linux estándar: /mnt/repositorio/...
  if (rutaNormalizada.startsWith('/mnt/repositorio')) {
    let windowsPath = rutaNormalizada.replace('/mnt/repositorio', '\\\\repositorio\\DirGesCat')
    windowsPath = windowsPath.replace(/\//g, '\\')
    return windowsPath
  }

  // Caso 2: Ruta que ya tiene formato Windows parcial con / : //repositorio/DirGesCat/...
  if (rutaNormalizada.toLowerCase().includes('//repositorio') ||
      rutaNormalizada.toLowerCase().includes('/repositorio/dirgescat')) {
    // Normalizar dobles barras al inicio
    let windowsPath = rutaNormalizada.replace('//', '\\\\')
    // Convertir el resto de / a \
    windowsPath = windowsPath.replace(/\//g, '\\')
    return windowsPath
  }

  // Caso 3: Ruta que ya está en formato Windows correcto: \\repositorio\DirGesCat\...
  if (rutaNormalizada.startsWith('\\\\repositorio') || rutaNormalizada.startsWith('\\repositorio')) {
    // Ya está correcta, solo asegurar barras invertidas
    return rutaNormalizada.replace(/\//g, '\\')
  }

  // Caso 4: Mensajes especiales o rutas no reconocidas
  // (podría ser "Sin ruta", "Sin directorio", etc.)
  return linuxPath
}

/**
 * Convierte una ruta de Windows UNC a formato Linux.
 *
 * Conversión:
 *   FROM: \\repositorio\DirGesCat\2510SP\...
 *   TO:   /mnt/repositorio/2510SP/...
 *
 * @param windowsPath - Ruta en formato Windows UNC
 * @returns Ruta en formato Linux o el valor original si no se puede convertir
 *
 * @example
 * windowsToLinuxPath('\\\\repositorio\\DirGesCat\\2510SP\\H_Informacion_Consulta')
 * // => '/mnt/repositorio/2510SP/H_Informacion_Consulta'
 */
export function windowsToLinuxPath(windowsPath: string | null | undefined): string | null | undefined {
  // Si es null, undefined o vacío, retornar tal cual
  if (!windowsPath) {
    return windowsPath
  }

  // Si no es una cadena válida, retornar tal cual
  if (typeof windowsPath !== 'string') {
    return windowsPath
  }

  // Normalizar la ruta (puede venir con \ o con \\)
  let normalizedPath = windowsPath.replace(/\\/g, '/')

  // Si no contiene el patrón de repositorio, retornar tal cual
  if (!normalizedPath.toLowerCase().includes('repositorio')) {
    return windowsPath
  }

  // Convertir de Windows UNC a Linux
  // //repositorio/DirGesCat/2510SP/... -> /mnt/repositorio/2510SP/...
  let linuxPath = normalizedPath.replace('//repositorio/DirGesCat/', '/mnt/repositorio/')
  linuxPath = linuxPath.replace('\\\\repositorio\\DirGesCat\\', '/mnt/repositorio/')

  return linuxPath
}

/**
 * Detecta si una ruta es de tipo Linux (comienza con /)
 *
 * @param path - Ruta a verificar
 * @returns true si es una ruta Linux
 */
export function isLinuxPath(path: string | null | undefined): boolean {
  if (!path || typeof path !== 'string') {
    return false
  }
  return path.trim().startsWith('/')
}

/**
 * Detecta si una ruta es de tipo Windows (comienza con \\ o letra de unidad)
 *
 * @param path - Ruta a verificar
 * @returns true si es una ruta Windows
 */
export function isWindowsPath(path: string | null | undefined): boolean {
  if (!path || typeof path !== 'string') {
    return false
  }
  const trimmed = path.trim()
  return trimmed.startsWith('\\\\') || /^[a-zA-Z]:/.test(trimmed)
}

/**
 * Detecta si una cadena parece ser una ruta de archivo/directorio
 *
 * @param value - Valor a verificar
 * @returns true si parece ser una ruta
 */
export function isPathLike(value: string | null | undefined): boolean {
  if (!value || typeof value !== 'string') {
    return false
  }
  const trimmed = value.trim()

  // Verificar patrones comunes de rutas
  return (
    trimmed.startsWith('/mnt/') ||
    trimmed.startsWith('\\\\') ||
    trimmed.startsWith('//') ||
    /^[a-zA-Z]:[\\/]/.test(trimmed) ||
    trimmed.toLowerCase().includes('repositorio')
  )
}

/**
 * Campos comunes que contienen rutas en los datos
 */
export const PATH_FIELDS = [
  'path',
  'path_file',
  'path_directorio',
  'ruta',
  'ruta_completa',
  'directorio',
  'ruta_directorio',
  'ruta_archivo',
  'ruta_acceso'
]

/**
 * Convierte todas las rutas en un objeto de datos para mostrar al usuario.
 *
 * Busca campos comunes de rutas y los convierte de Linux a Windows.
 *
 * @param dataDict - Objeto con datos que pueden contener rutas
 * @returns Objeto con rutas convertidas a formato Windows
 *
 * @example
 * const data = { path_file: '/mnt/repositorio/2510SP/file.pdf', nombre: 'Test' }
 * convertPathsForDisplay(data)
 * // => { path_file: '\\\\repositorio\\DirGesCat\\2510SP\\file.pdf', nombre: 'Test' }
 */
export function convertPathsForDisplay(dataDict: Record<string, any>): Record<string, any> {
  if (!dataDict || typeof dataDict !== 'object') {
    return dataDict
  }

  const result = { ...dataDict }

  for (const field of PATH_FIELDS) {
    if (field in result && result[field]) {
      result[field] = linuxToWindowsPath(result[field])
    }
  }

  return result
}

/**
 * Copia una ruta al portapapeles
 *
 * @param path - Ruta a copiar
 * @returns Promise que se resuelve cuando se copia exitosamente
 */
export async function copyPathToClipboard(path: string): Promise<void> {
  try {
    await navigator.clipboard.writeText(path)
  } catch (err) {
    // Fallback para navegadores antiguos
    const textArea = document.createElement('textarea')
    textArea.value = path
    textArea.style.position = 'fixed'
    textArea.style.left = '-999999px'
    textArea.style.top = '-999999px'
    document.body.appendChild(textArea)
    textArea.focus()
    textArea.select()
    document.execCommand('copy')
    textArea.remove()
  }
}

export default {
  linuxToWindowsPath,
  windowsToLinuxPath,
  isLinuxPath,
  isWindowsPath,
  isPathLike,
  convertPathsForDisplay,
  copyPathToClipboard,
  PATH_FIELDS
}

File: backend/backend/path_utils.py
"""
Utilidades para conversión de rutas entre formatos Linux y Windows.

Este módulo proporciona funciones para convertir rutas entre el formato Linux
usado en el servidor y el formato Windows UNC necesario para que los usuarios
puedan acceder a los archivos desde Excel.
"""


def linux_to_windows_path(linux_path):
    """
    Convierte una ruta de Linux a formato Windows UNC.

    Conversión:
        FROM: /mnt/repositorio/2510SP/...
        TO:   \\repositorio\DirGesCat\2510SP\...

    Args:
        linux_path (str): Ruta en formato Linux (ej: /mnt/repositorio/2510SP/...)

    Returns:
        str: Ruta en formato Windows UNC o el valor original si no se puede convertir

    Examples:
        >>> linux_to_windows_path('/mnt/repositorio/2510SP/H_Informacion_Consulta')
        '\\\\repositorio\\DirGesCat\\2510SP\\H_Informacion_Consulta'

        >>> linux_to_windows_path('Sin ruta')
        'Sin ruta'

        >>> linux_to_windows_path(None)
        None
    """
    # Si es None o vacío, retornar tal cual
    if not linux_path:
        return linux_path

    # Si no es una ruta válida, retornar tal cual
    if not isinstance(linux_path, str):
        return linux_path

    # Normalizar la ruta para procesar
    ruta_normalizada = linux_path.strip()

    # Caso 1: Ruta Linux estándar: /mnt/repositorio/...
    if ruta_normalizada.startswith('/mnt/repositorio'):
        windows_path = ruta_normalizada.replace('/mnt/repositorio', r'\\repositorio\DirGesCat')
        windows_path = windows_path.replace('/', '\\')
        return windows_path

    # Caso 2: Ruta que ya tiene formato Windows parcial con / : //repositorio/DirGesCat/...
    if '//repositorio' in ruta_normalizada.lower() or '/repositorio/dirgescat' in ruta_normalizada.lower():
        # Normalizar dobles barras al inicio
        windows_path = ruta_normalizada.replace('//', '\\\\', 1)
        # Convertir el resto de / a \
        windows_path = windows_path.replace('/', '\\')
        return windows_path

    # Caso 3: Ruta que ya está en formato Windows correcto: \\repositorio\DirGesCat\...
    if ruta_normalizada.startswith('\\\\repositorio') or ruta_normalizada.startswith(r'\\repositorio'):
        # Ya está correcta, solo asegurar barras invertidas
        return ruta_normalizada.replace('/', '\\')

    # Caso 4: Mensajes especiales o rutas no reconocidas
    # (podría ser "Sin ruta", "Sin directorio", etc.)
    return linux_path


def windows_to_linux_path(windows_path):
    """
    Convierte una ruta de Windows UNC a formato Linux.

    Conversión:
        FROM: \\repositorio\DirGesCat\2510SP\...
        TO:   /mnt/repositorio/2510SP/...

    Args:
        windows_path (str): Ruta en formato Windows UNC

    Returns:
        str: Ruta en formato Linux o el valor original si no se puede convertir

    Examples:
        >>> windows_to_linux_path('\\\\repositorio\\DirGesCat\\2510SP\\H_Informacion_Consulta')
        '/mnt/repositorio/2510SP/H_Informacion_Consulta'
    """
    # Si es None o vacío, retornar tal cual
    if not windows_path:
        return windows_path

    # Si no es una ruta válida, retornar tal cual
    if not isinstance(windows_path, str):
        return windows_path

    # Si ya es una ruta Linux válida, retornarla tal cual
    if windows_path.startswith('/mnt/repositorio'):
        return windows_path

    # Normalizar la ruta: convertir todas las barras invertidas a barras normales
    normalized_path = windows_path.replace('\\', '/')

    # Si no contiene el patrón de repositorio, retornar tal cual
    if 'repositorio' not in normalized_path.lower():
        return windows_path

    # Eliminar barras iniciales extras para normalizar
    # "//repositorio/..." o "/repositorio/..." -> "repositorio/..."
    while normalized_path.startswith('/'):
        normalized_path = normalized_path[1:]

    # Ahora la ruta debería ser: repositorio/DirGesCat/2510SP/...
    # Convertir a formato Linux: /mnt/repositorio/2510SP/...

    # Caso 1: Ruta con DirGesCat (formato estándar de red Windows)
    if normalized_path.lower().startswith('repositorio/dirgescat/'):
        # Quitar "repositorio/DirGesCat/" y agregar "/mnt/repositorio/"
        resto = normalized_path[len('repositorio/DirGesCat/'):]
        linux_path = '/mnt/repositorio/' + resto
        return linux_path

    # Caso 2: Ruta sin DirGesCat (por si ya viene parcialmente convertida)
    if normalized_path.lower().startswith('repositorio/'):
        resto = normalized_path[len('repositorio/'):]
        linux_path = '/mnt/repositorio/' + resto
        return linux_path

    # Si no coincide con ningún patrón conocido, retornar original
    return windows_path


def convert_paths_for_excel_export(data_dict):
    """
    Convierte todas las rutas en un diccionario de datos para exportación a Excel.

    Busca campos comunes de rutas y los convierte de Linux a Windows.

    Args:
        data_dict (dict): Diccionario con datos que pueden contener rutas

    Returns:
        dict: Diccionario con rutas convertidas a formato Windows

    Examples:
        >>> data = {'path_file': '/mnt/repositorio/2510SP/file.pdf', 'nombre': 'Test'}
        >>> convert_paths_for_excel_export(data)
        {'path_file': '\\\\repositorio\\DirGesCat\\2510SP\\file.pdf', 'nombre': 'Test'}
    """
    if not isinstance(data_dict, dict):
        return data_dict

    # Campos comunes que contienen rutas
    path_fields = [
        'path', 'path_file', 'path_directorio', 'ruta', 'ruta_completa',
        'directorio', 'ruta_directorio', 'ruta_archivo'
    ]

    result = data_dict.copy()

    for field in path_fields:
        if field in result and result[field]:
            result[field] = linux_to_windows_path(result[field])

    return result
